lib/galaxy/model/database_utils.py

```diff
@@ -0,0 +1,95 @@
+import sqlite3
+from contextlib import contextmanager
+
+from sqlalchemy import create_engine
+from sqlalchemy.engine.url import make_url
+from sqlalchemy.sql.compiler import IdentifierPreparer
+from sqlalchemy.sql.expression import text
+
+from galaxy.exceptions import ConfigurationError
+
+
+def database_exists(db_url, database=None):
+    """Check if database exists; connect with db_url.
+
+    If database is None, use the database name from db_url.
+    """
+    dbm = DatabaseManager.make_manager(db_url, database)
+    return dbm.exists()
+
+
+def create_database(db_url, database=None, encoding='utf8', template=None):
+    """Create database; connect with db_url.
+
+    If database is None, use the database name from db_url.
+    """
+    dbm = DatabaseManager.make_manager(db_url, database)
+    dbm.create(encoding, template)
+
+
+@contextmanager
+def sqlalchemy_engine(url):
+    engine = create_engine(url)
+    try:
+        yield engine
+    finally:
+        engine.dispose()
+
+
+class DatabaseManager:
+
+    @staticmethod
+    def make_manager(db_url, database):
+        if db_url.startswith('postgres'):
+            return PosgresDatabaseManager(db_url, database)
+        elif db_url.startswith('sqlite'):
+            return SqliteDatabaseManager(db_url, database)
+        else:
+            raise ConfigurationError(f'Invalid database URL: {db_url}')
+
+    def __init__(self, db_url, database):
+        self.url = make_url(db_url)
+        self.database = database
+        if not database:
+            self._handle_no_database()
+
+
+class PosgresDatabaseManager(DatabaseManager):
+
+    def _handle_no_database(self):
+        self.database = self.url.database  # use database from db_url
+        self.url.database = 'postgres'
```

Comment:
This won't work on some configuration, see https://github.com/kvesteri/sqlalchemy-utils/pull/372

Reply:
This method is called when no database is provided, so the assumption is that the supplied URL contains **both** the database to connect to AND the database to manipulate or check for existence. So, this implementation actually solves the issue discussed (the one you mention), as well as here https://github.com/kvesteri/sqlalchemy-utils/pull/463 - by changing the signature of create/exists functions to optionally include a database name if it different from the one that is used for authentication upon connection. This solution was briefly discussed in 463 (https://github.com/kvesteri/sqlalchemy-utils/pull/463#issuecomment-668214945), but not implemented as there was a preference to keep the signature unchanged (for several good reasons). We are not a library, so we only have to accommodate the needs of one application here - so we can make this change now (thus fixing an issue raised by @bernt-matthias  almost 2 years ago).